cng/ecdh.go

```diff
@@ -20,36 +20,31 @@ var errInvalidPrivateKey = errors.New("cng: invalid private key")
 
 type ecdhAlgorithm struct {
 	handle bcrypt.ALG_HANDLE
+	bits   uint32
 }
 
-func loadECDH(curve string) (h ecdhAlgorithm, bits uint32, err error) {
-	var id string
-	switch curve {
-	case "P-256":
-		id, bits = bcrypt.ECC_CURVE_NISTP256, 256
-	case "P-384":
-		id, bits = bcrypt.ECC_CURVE_NISTP384, 384
-	case "P-521":
-		id, bits = bcrypt.ECC_CURVE_NISTP521, 521
-	case "X25519":
-		id, bits = bcrypt.ECC_CURVE_25519, 255
-	default:
-		err = errUnknownCurve
-	}
-	if err != nil {
-		return
-	}
-	v, err := loadOrStoreAlg(bcrypt.ECDH_ALGORITHM, bcrypt.ALG_NONE_FLAG, id, func(h bcrypt.ALG_HANDLE) (interface{}, error) {
-		err := setString(bcrypt.HANDLE(h), bcrypt.ECC_CURVE_NAME, id)
+func loadECDH(curve string) (ecdhAlgorithm, error) {
+	return loadOrStoreAlg(bcrypt.ECDH_ALGORITHM, bcrypt.ALG_NONE_FLAG, curve, func(h bcrypt.ALG_HANDLE) (ecdhAlgorithm, error) {
+		var name string
+		var bits uint32
+		switch curve {
+		case "P-256":
```

Comment:
Looks like these values are also used in other files - would it be idiomatic to extract to some sort of const or enum? that multiple files could use?

Reply:
There is no general rule for when to use raw string literals rather than const or enum variables. Here I'm following the Go toolchain and standard library unwritten convention: always use raw string literals even if the string is repeated one thousand times. Specifically, the standard library always hardcode curve names. The exception is when you have to interop with an external library, such as bcrypt, in which case it is common to have a single package that encapsulate all constants and functions.

This approach could lead to bugs if a type slips in, but it would be cached by unit tests anyway. The benefit is that the code is much readable (thus easier to maintain), but that's maybe because I'm used to it. See for example this switch case, there is more noise in the first two statements than in the last two:

```go
switch curve {
case curveNameP256:
	name, bits = bcrypt.ECC_CURVE_NISTP256, 256
case curveNameP384:
	name, bits = bcrypt.ECC_CURVE_NISTP384, 384
case "P-521":
	name, bits = bcrypt.ECC_CURVE_NISTP521, 521
case "X25519":
	name, bits = bcrypt.ECC_CURVE_25519, 255
default:
	return ecdhAlgorithm{}, errUnknownCurve
}
```